imports/plugins/core/core/server/Reaction/addRolesToGroups.js

```diff
@@ -77,17 +29,92 @@ export default function addRolesToGroups(options = { allShops: false, roles: [],
     }
   };
-  const multi = { multi: true };
 
   if (!allShops) {
-    // if we're not updating for all shops, we should only update for the shops passed in.
+    // If we're not updating for all shops, we should only update for the shops passed in.
     query.shopId = { $in: shops || [] };
     Logger.debug(`Adding Roles: ${roles} to Groups: ${groups} for shops: ${shops}`);
   } else {
     Logger.debug(`Adding Roles: ${roles} to Groups: ${groups} for all shops`);
   }
 
-  addRolesToUsersInGroups({ query, roles });
+  // Check if each group already has the roles. If so, skip updates for group & its users
+  Groups.find(query).forEach((group) => {
+    let areRolesInGroup = true;
+    roles.forEach((role) => {
+      if (group.permissions.includes(role) === false) {
+        Logger.debug(`Role ${role} is not in ${group.name} group`);
+        areRolesInGroup = false;
+      }
+    });
+
+    if (areRolesInGroup === false) {
+      Logger.debug(`Adding roles (${roles}) to ${group.name} group (${group._id})`);
+      // Run heaviest queries asynchronously in the background
+      Meteor.defer(() => {
+        addRolesToGroupAndUsers(group, roles);
+      });
+    } else {
+      Logger.debug(`Skipping roles already assigned to ${group.name} group and its users`);
+    }
+  });
+}
+
+/**
+ * @name addRolesToGroupAndUsers
+ * @private
+ * @summary Adds the given roles to the given group, and updates users if necessary
+ * @param {Object} group - Group to add roles to, and update users for
+ * @param {String} group._id - Group's _id
+ * @param {String} group.shopId - _id of shop group belongs to
+ * @param {String} group.name - Name of group
+ * @param {Array} roles - Array of roles/permissions (strings)
+ * @returns {undefined} Nothing
+ */
+function addRolesToGroupAndUsers({ _id, shopId, name }, roles) {
+  Groups.update({ _id }, { $addToSet: { permissions: { $each: roles } } });
+
+  const maxAccountsPerUpdate = 100000;
+  const accountSelector = { groups: _id };
+  const numAccounts = Accounts.find(accountSelector).count();
+  if (numAccounts === 0) {
+    Logger.debug("No users need roles updated");
+    return;
+  }
+
+  Logger.debug(`Number of users to add roles to: ${numAccounts}`);
+  const numQueriesNeeded = Math.ceil(numAccounts / maxAccountsPerUpdate);
+  Logger.debug(`Number of updates needed: ${numQueriesNeeded}`);
+
+  const accountOptions = {
+    fields: { _id: 1 },
+    sort: { _id: 1 },
+    limit: maxAccountsPerUpdate
+  };
+  let lastUserIdUpdated = "";
+
+  for (let inc = 0; inc < numQueriesNeeded; inc += 1) {
+    Logger.debug(`Processing user role update #${inc + 1} of ${numQueriesNeeded} for ${name} group`);
+
+    if (lastUserIdUpdated) {
+      accountSelector._id = {
+        $gt: lastUserIdUpdated
+      };
+    }
+
+    const userIds = Accounts.find(accountSelector, accountOptions).map((account) => account._id);
+    lastUserIdUpdated = userIds[userIds.length - 1];
+
+    const selector = { _id: { $in: userIds } };
```

Comment:
@dancastellon Since we already have the range of `userIds` we want to update, wouldn't it make more sense to send a `in` filter in the `selector`? Something like `{ _id: { $gte: userIds[0], $lte: lastUserIdUpdated]} }`
Since the range is on a indexed field it will be executed pretty quickly and we won't have to send 100k values.

Reply:
@Akarshit done, great catch!